guest-js/index.ts

```diff
@@ -97,101 +121,215 @@ export class DownloadEventManager {
-   public url: string;
-   public path: string;
-   public progress: number;
-   public state: DownloadState;
+abstract class DownloadBase implements DownloadItem {
+   public readonly key: string;
+   public readonly url: string;
+   public readonly path: string;
+   public readonly progress: number;
+   public abstract readonly state: DownloadState;
 
-   public constructor(item: DownloadItem) {
+   protected constructor(item: DownloadItem) {
       this.key = item.key;
       this.url = item.url;
       this.path = item.path;
       this.progress = item.progress;
-      this.state = item.state;
    }
 
    /**
-   * Starts the download.
+   * Listen for changes to the download.
+   * To avoid memory leaks, the `unlisten` function returned by the promise
+   * should be called when no longer required.
+   * @param onChanged - Callback function invoked when the download has changed.
+   * @returns A promise with a function to remove the download listener.
+   *
+   * @example
+   * ```ts
+   * const unlisten = await download.listen((updatedDownload) => {
+   *   console.log('Download:', updatedDownload);
+   *   if (updatedDownload.state === DownloadState.PAUSED) {
+   *     updatedDownload.resume(); // TypeScript knows this is valid
+   *   }
+   * });
+   *
+   * // To stop listening
+   * unlisten();
+   * ```
+   */
+   public async listen(listener: (download: Download) => void): Promise<UnlistenFn> {
+      return DownloadEventManager.shared.addListener(this.key, listener);
+   }
+}
+
+/**
+ * A download that has been cancelled.
+ * Terminal state - no further actions available.
+ */
+export class CancelledDownload extends DownloadBase {
+   public readonly state = DownloadState.CANCELLED;
+
+   public constructor(item: DownloadItem) {
+      super(item);
+   }
+}
+
+/**
+ * A download that has completed successfully.
+ * Terminal state - no further actions available.
+ */
+export class CompletedDownload extends DownloadBase {
+   public readonly state = DownloadState.COMPLETED;
+
+   public constructor(item: DownloadItem) {
+      super(item);
+   }
+}
+
+/**
+ * A download in an unknown state.
+ * This may occur if the plugin returns an unrecognized state.
+ */
+export class UnknownDownload extends DownloadBase {
+   public readonly state = DownloadState.UNKNOWN;
+
+   public constructor(item: DownloadItem) {
+      super(item);
+   }
+}
+
+/**
+ * A download that has been paused.
+ * Can be resumed or cancelled.
+ */
+export class PausedDownload extends DownloadBase {
+   public readonly state = DownloadState.PAUSED;
+
+   public constructor(item: DownloadItem) {
+      super(item);
+   }
+
+   /**
+   * Resumes the download.
    * @returns A promise with the updated download.
    */
-   public async start(): Promise<Download> {
-      return new Download(await invoke('plugin:download|start', { key: this.key }));
+   // eslint-disable-next-line @typescript-eslint/no-use-before-define
+   public async resume(): Promise<ActiveDownload> {
```

Comment:
ADIRL: The IPC will respond with some kind of `response` objects with a few different helpful properties. We would return that response object for each of these methods.

One of the response object's properties would be a `Download` object that represents the _actual_ state at the time the Rust layer sends its response. In most cases, this would be a download object with the state you'd expect given the action you just took, e.g. for the `resume` method a `Download` object with `status` of `InProgress`. But it also allows us to return some other state in the case when the action we tried to take doesn't match up with the actual backend-state of the download, e.g. sending a "pause" action when the download is in a "Completed" state.